___
# Atividade: <br> <center> Ciência de Dados para Engenharia: <br> <center> O que são "dados"? O que queremos fazer com eles? 
___

## Aula 00

**Objetivo da aula:** ao fim desta aula, o aluno será capaz de: <br>
 * compreender a necessidade de se ter um objetivo (ou hipótese teórica) previamente definido com intuito de guiá-lo na manipulação de uma base de dados; e 
 * realizar algumas manipulações em um *dataframe* com auxílio do **Pandas**.


___
<div id="indice"></div>

## Índice

1. [Hipóteses teóricas e conjunto de dados](#hipoteses)
    - [Exemplos de hipóteses teóricas em algumas áreas](#exemplosht)
    - [Exercício 1](#ex1)
    
    
2. [Glossário para base de dados: `WorldBank.xlsx`](#dados)
    - [Algumas hipóteses teóricas:](#htwb)
    - [Exercício 2](#ex2)


3. [Manipulação de base de dados](#manipulacao)
    - [Abrindo um arquivo e visualizando seu conteúdo](#abrindo-arquivo)
    - [Visualizando primeiras linhas de uma base de dados](#head)
    - [Visualizando últimas linhas de uma base de dados](#tail)
    - [Exercício 3](#ex3)


___
<div id="hipoteses"></div>

## 1. Hipóteses e conjunto de dados

Em uma análise exploratória de dados, devemos partir de **hipóteses teóricas** que nos guiam na seleção das variáveis relevantes a serem incorporadas em um conjunto de dados e também direcionam como essas devem ser exploradas para embasar as tomadas de decisão relacionadas ao problema em questão.

<div id="exemplosht"></div>

### Exemplos de hipóteses teóricas em algumas áreas:


 * **Eficácia de uma Nova Droga:** A nova droga X é eficaz no tratamento da doença Y em comparação com uma terapia padrão? <br>
   *Dados de ensaios clínicos e resultados de pacientes podem ser analisados para determinar a eficácia da nova droga*;
    <p>

 * **Relação Entre Fatores de Estilo de Vida e Doenças Cardiovasculares:** Existe uma relação entre o consumo de tabaco, dieta pouco saudável e a incidência de doenças cardíacas? <br>
   *Dados de estudos epidemiológicos podem ser usados para identificar associações*
     <p>

 * **Impacto do Aumento do Salário Mínimo:** O aumento do salário mínimo leva a uma redução na taxa de desemprego? <br>
   *Dados de emprego e salário podem ser analisados para avaliar o impacto econômico dessa política*;
   <p>

 * **Análise de Tendências de Consumo:** Como as preferências dos consumidores estão mudando ao longo do tempo e qual é o impacto nas vendas? <br>
   *Dados de vendas e comportamento do consumidor podem ser usados para identificar tendências e ajustar estratégias de negócios*;

 * **Previsão de Falhas em Máquinas Industriais:** É possível prever falhas em equipamentos industriais com base em padrões de dados de sensores? <br>
   *Dados de sensores podem ser analisados para desenvolver modelos de previsão de falha*;
   <p>

 * **Otimização de Rotas de Transporte:** Como otimizar rotas de entrega para minimizar custos e tempo? <br>
   *Dados de tráfego, condições climáticas e logística podem ser usados para criar algoritmos de otimização de rotas*;
   <p>

 * **Monitoramento de Qualidade do Ar:** Como a poluição do ar afeta a saúde das populações urbanas? <br>
   *Dados de qualidade do ar e dados de saúde pública podem ser analisados para identificar possíveis correlações*;
   <p>

 * **Aquecimento Global e Níveis do Mar:** Existe uma relação direta entre as emissões de gases de efeito estufa e o aumento dos níveis do mar? <br>
   *Dados climáticos e de medição de níveis do mar podem ser utilizados para avaliar essa relação*.
   <p>

<br>

Lembre-se de que esses exemplos são hipóteses que são frequentemente questionadas em problemas reais e que a análise exploratória de dados é uma ferramenta poderosa para buscar respostas e *insights* nessas áreas.

[Voltar ao índice](#indice)

<div id="ex1"></div>

## Exercício 1: 

Para ilustrar, proponha um cenário de interesse particular e desenvolva hipóteses teóricas potenciais. <br>
Em seguida, identifique variáveis relevantes (sejam qualitativas ou quantitativas) que seriam essenciais em sua base de dados, visando a condução de uma análise exploratória de dados abrangente para tomada de decisão. <br>


[Voltar ao índice](#indice)

___
<div id="dados"></div>


## 2. Glossário para base de dados: `WorldBank.xlsx`

Nesta atividade, será utilizado o arquivo `WorldBank.xlsx`. <br>
A seguir, são apresentadas as variáveis e seus significados para identificar o que cada uma das colunas dessa base de dados significa:

- **Country**: nome do país;
- **Population**: tamanho populacional, ou seja, número total de indivíduos da população;
- **GDPcapita**: produto interno bruto (PIB) per capita;
- **surface**: área total do país, em km$^2$;
- **region**: nome da região que país faz parte, sendo categorizada em: africa_north, africa_sub_saharan, america_north, america_south, asia_west, east_asia_pacific, europe_east e europe_west; e
- **landlocked**: indica se país não tem fronteira com litoral, sendo categorizada em: 0=com litoral; e 1=sem litoral.

<div id="htwb"></div>

### Algumas hipóteses teóricas:

A seguir, citamos algumas hipóteses teóricas que poderiam ser validadas com uso da base nos dados fornecida.

 * **Hipótese: O PIB per capita está correlacionado com a população de um país?** <br>
 Essa hipótese questiona se países com populações maiores tendem a ter um PIB per capita mais alto ou mais baixo.
 <p>
    
 * **Hipótese: Países sem litoral têm um PIB per capita mais baixo em comparação com os países com litoral?** <br>
 Essa hipótese explora se a presença de litoral está relacionada a um PIB per capita mais alto devido a potenciais vantagens econômicas.
 <p>

 * **Hipótese: A área total de um país está correlacionada com sua população?** <br>
 Esta hipótese considera se países maiores em área tendem a ter uma população maior ou menor.
 <p>
    
 * **Hipótese: Certas regiões geográficas têm uma probabilidade maior de serem países sem litoral?** <br>
 Esta hipótese explora se determinadas regiões geográficas têm uma tendência maior de conter países sem litoral.
 <p>
    
 * **Hipótese: Países com PIB per capita mais alto estão localizados predominantemente em regiões específicas?** <br>
 Aqui, você está explorando se certas regiões geográficas têm uma tendência maior de abrigar países com PIB per capita mais alto.
 <p>
    
Essas são apenas algumas hipóteses que você pode explorar com base nos dados fornecidos. <br>
Lembre-se de que, na prática, uma análise exploratória de dados envolveria a coleta, limpeza e análise dos dados para validar ou refutar essas hipóteses.

<div id="ex2"></div>

## Exercício 2: 

Considerando os dados apresentados no arquivo `WorldBank.xlsx`, apresente outras hipóteses teóricas que poderiam ser interessantes dentro de um contexto prático. <br>
Ainda, discuta se seria interessante ter mais informações (colunas) na base de dados. Se sim, quais informações seriam essas?

[Voltar ao índice](#indice)

___
<div id="manipulacao"></div>

## 3. Manipulação da base de dados com **Pandas**

A biblioteca [**Pandas**](https://pandas.pydata.org/docs/user_guide/index.html) é um conjunto de funções para o Python com intuito de trabalhar com *Data Science*. <br>
Essa biblioteca permite, além de abrir diversos tipos de arquivos, trabalhar também com o ferramental descritivo para responder perguntas sobre seus dados.

In [1]:
# Importando as bibliotecas necessárias para esta atividade
import pandas as pd

<div id="abrindo-arquivo"></div>

### Abrindo um arquivo e visualizando seu conteúdo:

Vamos começar abrindo um arquivo no formato Excel. <br>
Mas vamos verificar ANTES se o arquivo `WorldBank.xlsx` foi salvo na mesma pasta em que salvou este arquivo notebook. <br>
O comando a seguir importa a biblioteca [**os**](https://docs.python.org/pt-br/3/library/os.html) e mostra a pasta de trabalho!

In [2]:
# Importando biblioteca para Sistema Operacional
import os

In [3]:
print(f'Esperamos trabalhar no diretório: \n{os.getcwd()}\n')

Esperamos trabalhar no diretório: 
/home/vitorhpk/www/python/cdados



In [4]:
filename = 'WorldBank.xlsx'

if filename in os.listdir():
    print(f'Parece que o arquivo {filename} está na mesma pasta do notebook, yay!')
    
else:
    print(f'Não encontrei o arquivo {filename}.\n'
          f'Esse deve ser salvo na mesma pasta em que se encontrar este arquivo jupyter notebook')

Parece que o arquivo WorldBank.xlsx está na mesma pasta do notebook, yay!


**ATENÇÃO:** <br>
Prosseguir nesta atividade apenas se a mensagem `Parece que o arquivo {filename} está na mesma pasta do notebook, yay!'` aparecer ao código acima.<br>

...

A seguir, o arquivo `WorldBank.xlsx` será aberto com uso da biblioteca `pandas` e seu conteúdo será armazenado em uma variável nomeada `dados`.

In [5]:
dados = pd.read_excel('WorldBank.xlsx')

# Se esse comando não funcionar, pode ser que você não tenha instalado a 
# biblioteca 'xlrd', leia a mensagem de erro até o fim.
# tire o # da linha a seguir caso tenha erro ao executar esta célula
# !pip install xlrd

Vamos ver quais informações estão disponíveis em `dados`:

In [6]:
type(dados)

pandas.core.frame.DataFrame

In [7]:
dados.shape

(168, 6)

In [8]:
dados.dtypes

Country        object
Population      int64
GDPcapita     float64
surface         int64
region         object
landlocked      int64
dtype: object

...

Com os *outputs* acima, notamos que `dados` é um objeto do tipo `DataFrame`, o tipo padrão usado no pandas para representar uma base de dados. <br> 
É um *dataframe* com 168 linhas e 6 colunas. <br>
Por fim, é possível identificar qual é o tipo de informação armazenado em cada coluna, como, por exemplo, `Population` é número inteiro e `GDPcapita` é contém informação numérica mas com casas decimais.

A seguir, vamos ver parte do que está contido em `dados`:

In [9]:
dados

Country  Population     GDPcapita  surface  \
0                Albania     2901883   1915.424459    28750   
1                Algeria    36036159   2231.980246  2381740   
2                 Angola    21219954    623.245275  1246700   
3    Antigua and Barbuda       87233  10614.794315      440   
4              Argentina    41222875  10749.319224  2780400   
..                   ...         ...           ...      ...   
163              Vanuatu      236299   1522.384040    12190   
164            Venezuela    28995745   5528.363114   912050   
165              Vietnam    88357775    722.810053   331050   
166               Zambia    13917439    434.659804   752610   
167             Zimbabwe    13973897    322.556928   390760   

                 region  landlocked  
0           europe_east           0  
1          africa_north           0  
2    africa_sub_saharan           0  
3         america_north           0  
4         america_south           0  
..                  ...         ...  
163   east_asia_pacific           0  
164       america_south           0  
165   east_asia_pacific           0  
166  africa_sub_saharan           1  
167  africa_sub_saharan           1  

[168 rows x 6 columns]

Observando o *output* acima, notamos que não imprimiu todas as linhas do * dataframe*. <br>
Isso acontece porque o conjunto de dados pode ser **muito** grande. <br>

Quando queremos apenas ter uma ideia de quais informações estão disponíveis no *dataframe*, então pode ser interessante visualizar poucas linhas. A seguir, será apresentado como fazer isso.

[Voltar ao índice](#indice)

<div id="head"></div>

### Visualizando primeiras linhas de uma base de dados

Para ver somente as primeiras linhas podemos usar o comando `.head()`:

In [10]:
dados.head()

Country  Population     GDPcapita  surface              region  \
0              Albania     2901883   1915.424459    28750         europe_east   
1              Algeria    36036159   2231.980246  2381740        africa_north   
2               Angola    21219954    623.245275  1246700  africa_sub_saharan   
3  Antigua and Barbuda       87233  10614.794315      440       america_north   
4            Argentina    41222875  10749.319224  2780400       america_south   

   landlocked  
0           0  
1           0  
2           0  
3           0  
4           0

Você pode especificar outra quantidade de linhas passando um número inteiro como argumento:

In [11]:
dados.head(10)

Country  Population     GDPcapita  surface              region  \
0              Albania     2901883   1915.424459    28750         europe_east   
1              Algeria    36036159   2231.980246  2381740        africa_north   
2               Angola    21219954    623.245275  1246700  africa_sub_saharan   
3  Antigua and Barbuda       87233  10614.794315      440       america_north   
4            Argentina    41222875  10749.319224  2780400       america_south   
5              Armenia     2963496   1326.710864    29740         europe_east   
6            Australia    22162863  25190.839860  7741220   east_asia_pacific   
7              Austria     8391986  26642.993858    83870         europe_west   
8           Azerbaijan     9099893   2344.810935    86600         europe_east   
9              Bahamas      360830  19395.152312    13880       america_north   

   landlocked  
0           0  
1           0  
2           0  
3           0  
4           0  
5           1  
6           0  
7           1  
8           1  
9           0

[Voltar ao índice](#indice)

<div id="tail"></div>

### Visualizando últimas linhas de uma base de dados

Se preferir, podemos ver as últimas linhas com o comando `.tail()`, que também pode receber a quantidade de linhas como argumento:



In [12]:
dados.tail(3)

Country  Population   GDPcapita  surface              region  landlocked
165   Vietnam    88357775  722.810053   331050   east_asia_pacific           0
166    Zambia    13917439  434.659804   752610  africa_sub_saharan           1
167  Zimbabwe    13973897  322.556928   390760  africa_sub_saharan           1

[Voltar ao índice](#indice)

<div id="ex3"></div>

### Exercício 3:

Suponha os seguintes objetivos:
 * Construir uma nova base de dados **sem** as colunas `Country` e `surface` da base original `dados`.
 * Construir uma nova base de dados **sem** as linhas 1, 4 e 6 da base original `dados`.

Encontre como fazer isso pesquisando no seu navegador.

In [15]:
dados.drop(columns=['Country', 'surface'])

Population     GDPcapita              region  landlocked
0       2901883   1915.424459         europe_east           0
1      36036159   2231.980246        africa_north           0
2      21219954    623.245275  africa_sub_saharan           0
3         87233  10614.794315       america_north           0
4      41222875  10749.319224       america_south           0
..          ...           ...                 ...         ...
163      236299   1522.384040   east_asia_pacific           0
164    28995745   5528.363114       america_south           0
165    88357775    722.810053   east_asia_pacific           0
166    13917439    434.659804  africa_sub_saharan           1
167    13973897    322.556928  africa_sub_saharan           1

[168 rows x 4 columns]

In [16]:
dados.drop(index=[1,4,6])

Country  Population     GDPcapita  surface  \
0                Albania     2901883   1915.424459    28750   
2                 Angola    21219954    623.245275  1246700   
3    Antigua and Barbuda       87233  10614.794315      440   
5                Armenia     2963496   1326.710864    29740   
7                Austria     8391986  26642.993858    83870   
..                   ...         ...           ...      ...   
163              Vanuatu      236299   1522.384040    12190   
164            Venezuela    28995745   5528.363114   912050   
165              Vietnam    88357775    722.810053   331050   
166               Zambia    13917439    434.659804   752610   
167             Zimbabwe    13973897    322.556928   390760   

                 region  landlocked  
0           europe_east           0  
2    africa_sub_saharan           0  
3         america_north           0  
5           europe_east           1  
7           europe_west           1  
..                  ...         ...  
163   east_asia_pacific           0  
164       america_south           0  
165   east_asia_pacific           0  
166  africa_sub_saharan           1  
167  africa_sub_saharan           1  

[165 rows x 6 columns]

[Voltar ao índice](#indice)